File: library/tract-renderer/src/main/java/org/cru/godtools/tract/analytics/model/ContentAnalyticsActionEvent.java
package org.cru.godtools.tract.analytics.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.common.base.Strings;

import org.cru.godtools.analytics.AdobeAnalyticsService;
import org.cru.godtools.analytics.AnalyticsService;
import org.cru.godtools.analytics.model.AnalyticsActionEvent;
import org.cru.godtools.tract.model.AnalyticsEvent;

import java.util.Map;

public class ContentAnalyticsActionEvent extends AnalyticsActionEvent {
    @NonNull
    private final AnalyticsEvent mAnalyticsEvent;

    public ContentAnalyticsActionEvent(@NonNull final AnalyticsEvent event) {
        mAnalyticsEvent = event;
    }

    @Override
    public boolean trackInService(@NonNull final AnalyticsService service) {
        if (service instanceof AdobeAnalyticsService) {
            return mAnalyticsEvent.isForSystem(AnalyticsEvent.System.ADOBE);
        }
        return false;
    }

    @NonNull
    @Override
    public String getAction() {
        return Strings.nullToEmpty(mAnalyticsEvent.getAction());
    }

    @Nullable
    @Override
    public Map<String, ?> getAttributes() {
        return mAnalyticsEvent.getAttributes();
    }
}

File: ui/base-tool/src/main/java/org/cru/godtools/base/tool/analytics/model/ContentAnalyticsActionEvent.kt
package org.cru.godtools.base.tool.analytics.model

import android.os.Bundle
import androidx.annotation.VisibleForTesting
import org.cru.godtools.analytics.model.AnalyticsActionEvent
import org.cru.godtools.analytics.model.AnalyticsSystem
import org.cru.godtools.tool.model.AnalyticsEvent

class ContentAnalyticsActionEvent(@VisibleForTesting val event: AnalyticsEvent) :
    AnalyticsActionEvent(action = event.action.orEmpty()) {
    override fun isForSystem(system: AnalyticsSystem) = event.isForSystem(
        when (system) {
            AnalyticsSystem.ADOBE -> AnalyticsEvent.System.ADOBE
            AnalyticsSystem.APPSFLYER -> AnalyticsEvent.System.APPSFLYER
            AnalyticsSystem.FACEBOOK -> AnalyticsEvent.System.FACEBOOK
            AnalyticsSystem.FIREBASE -> AnalyticsEvent.System.FIREBASE
            AnalyticsSystem.SNOWPLOW -> AnalyticsEvent.System.SNOWPLOW
        }
    )

    override val appSection get() = event.manifest.code

    override val adobeAttributes get() = event.attributes

    override val firebaseEventName get() = when {
        isForSystem(AnalyticsSystem.ADOBE) -> event.action?.sanitizeAdobeNameForFirebase().orEmpty()
        else -> super.firebaseEventName
    }

    override val firebaseParams
        get() = Bundle().apply {
            when {
                event.isForSystem(AnalyticsEvent.System.FIREBASE) ->
                    event.attributes.forEach { putString(it.key, it.value) }
                event.isForSystem(AnalyticsEvent.System.ADOBE) ->
                    event.attributes.forEach { putString(it.key.sanitizeAdobeNameForFirebase(), it.value) }
            }
        }
}
